Initial Gitlab CI support [gitlab-ci-dev]

adrienbernede opened the pull request:
This PR introduces a first implementation of a Gitlab pipeline for MFEM.

In particular, this pipeline is setup to:
- perform some static checks (code style and documentation)
- build MFEM prerequisites
- build MFEM for different configurations
- perform sanity checks
- run baseline test and share the diff

TODO:
- [X] use existing style test (introduced by #1244)
- [x] validate baseline update mechanism and restrict it to master branch
- CANCELLED: fix jobs triggered when pipeline failed early. [Bug in Gitlab](https://gitlab.com/gitlab-org/gitlab/issues/31526)
- [x] fix concurrency issue on baseline test (mfem link naming): will solve itself by adding baseline test in mfem (no need to call tests from an outside repo)
- [x] print diff in the logs (style check and baseline check)

<!--GHEX{"id":1203,"author":"adrienbernede","editor":"tzanio","reviewers":["acfisher","tzanio"],"assignment":"2019-12-19","approval":"2020-03-18T21:31:01.892Z","merge":"2020-03-19T15:50:10.423Z"}XEHG--><!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1203](https://github.com/mfem/mfem/pull/1203) | @adrienbernede | @tzanio | @acfisher + @tzanio | 12/19/19 | 03/18/20 | 03/19/20 | |
<!--ELBATXEHG-->

adrienbernede commented:
LGTM

tzanio commented:
@acfisher and @v-dobrev, reminder that the reviews for this PR are due in a week

tzanio commented:
@mfem/reviewers -- this PR will affect the review process, so please take a look and provide your input to @adrienbernede and @acfisher.

barker29 commented:
I suggest updating `CONTRIBUTING.md` to give some basic info about how to use/check the Gitlab CI features.

adrienbernede commented:
@barker29 
I updated `CONTRIBUTING.md`, please let me know if that’s what you expected.

adrienbernede commented:
@tzanio @acfisher 

We are experiencing some issues with Gitlab mirroring.
Basically, since we filter out branches that are not "approved", Gitlab fails to create them and considers each failure as an error. Eventually, Gitlab blocks mirroring because of those recurring errors. This does not happen when the mirror repo has just been created. Which is why it appeared to be working in December.

In short, error handling in Gitlab is at fault here, and @tgmachina will be discussing possible improvements with the Gitlab team.

If this PR is accepted, I think we should still merge it. However the Gitlab CI would not be working until further notice.

In the meantime, we could create another MFEM repo on Gitlab (not a mirror repo), in which branches can be pushed to trigger the pipeline. This would be a manual process, but anyone with an LC Gitlab account could trigger tests when necessary. Scaling isn’t guaranteed though.

tgmachina commented:
@adrienbernede 

Currently, the result of that error handling is that mirroring for the repo stops--would it suffice right now to re-enable it periodically in GitLab? It's also not ideal, but maybe easier for one person to just turn it back on in lieu of maintaining another repo that is having CI triggered on it manually, just a thought.

I spoke with the GitLab team today and they agree that this is a priority. I'll open an issue with them and we'll work on getting the fix made in GitLab server. Meanwhile, I could work on patching the offending code myself in our own instance.

adrienbernede commented:
I first I thought that mirroring was paused if too many errors occurred during one mirroring attempt.
Umpire case proved me wrong. One error interrupts the mirroring. After several interruptions, the mirroring is paused.
Re-enabling mirroring would NOT make it to work again, even for a little while. I was also confused by the false positive we sometimes get (mirroring marked as successful in settings page, but marked as failing on project page).

tzanio commented on makefile:
```diff
@@ -675,6 +678,16 @@ style:
 	   echo "No source files were changed.";\
 	fi
 
+style-check:
```
I suggest we handle this by introducing `code-style` script in #1219 (see https://mybitbucket.llnl.gov/projects/MFEM/repos/tests/browse/code-style)

adrienbernede replied:
@tzanio @acfisher 
It seems to me that the test you mention actually applies astyle change.
However in this PR, @acfisher introduced an "astyle check" that tells if style is OK.

I would rather use this check, because I don’t think it would be easy to modify the code during the PR testing. However it’s the perfect place to tell the developers they should have check their code for style compliance.

Is that OK?

tzanio commented:
Delaying by a week to give folks time to address the above issue.

tzanio commented:
Delaying to give folks time to address issues.

tzanio commented:
Merged in `master` to trigger new Travis checks...

adrienbernede commented:
> Merged in `master` to trigger new Travis checks...

I was surprised to see "merged in master" in my emails...

adrienbernede commented:
LGTM

adrienbernede commented:
Let’s trigger tests on Gitlab too.

adrienbernede commented:
Now using the same script for style check as in Travis.

@acfisher, @tzanio suggested to revert the (now unused) change in the makefile introducing style-check target (308861c6d09cd8c9d910f64d423a84f72b5c4f75).
Sounds good to me, can you please confirm that’s OK for you too?

acfisher commented:
Yes, we now have another approach for that functionality.  Sounds good to me.

adrienbernede commented:
Done!

adrienbernede commented:
@acfisher @tzanio 
I’m not sure what we agreed on:
- Do you remove doc and style check from gitlab ci (sounds OK to me)
- Do we want to use the tests repo, or do we move baseline check to the new tests/scripts directory? If yes, maybe the tests repo can be converted to a baseline repo?

tzanio commented:
LGTM

tzanio commented:
This needs more testing and there still seem to be some issues... extending the review due date to 02/11/20.

acfisher commented:
LGTM

acfisher commented:
@adrienbernede it looks like the lib builds are failing?

adrienbernede commented:
LGTM

adrienbernede commented:
LGTM

adrienbernede commented:
LGTM

tzanio commented on tests/scripts/code-style:
```diff
@@ -48,8 +48,10 @@ shift
+
 make distclean
-make style
+make style ASTYLE_BIN=${ASTYLE:-astyle}
```
Why is this necessary?

adrienbernede replied:
When I was using astyle in Gitlab, I wanted to specify the full path to the desired version on astyle.
This is now useless and has been reverted.

tzanio commented on makefile:
```diff
@@ -662,7 +662,8 @@ status info:
-ASTYLE = astyle --options=$(SRC)config/mfem.astylerc
+ASTYLE_BIN = astyle
+ASTYLE = $(ASTYLE_BIN) --options=$(SRC)config/mfem.astylerc
 FORMAT_FILES = $(foreach dir,$(DIRS) $(EM_DIRS) config,"$(dir)/*.?pp")
```
This is OK in general, but why will the `astyle` binary have a different name?

adrienbernede replied:
When I was using astyle in Gitlab, I wanted to specify the full path to the desired version on astyle.
This is now useless and has been reverted.

tzanio commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,198 @@
+              export JOB_ID=$(squeue -h --name=${MFEM_ALLOC_NAME} --format=%A)
+              export EXEC_PREFIX="srun --jobid=${JOB_ID}"
+          fi
+        - $EXEC_PREFIX make -j 4 ${METIS} PREFIX=${BUILD_PATH}/ BUILD_DIR=${BUILD_PATH} LOG_DIR=${BUILD_PATH} CC=cc CXX=c++
```
Shouldn't the `-j 4` and the `-j 8` below be machine-dependent?

adrienbernede replied:
Now configurable per machine

tzanio commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,198 @@
+
+# The BUILD_ROOT is unique to the branch and the pipeline ID. Branch is not
+# necessary but makes it easier to retrieve a data. This means that there cannot
+# be concurrency between 2 pipelines.
```
I don't understand what is meant by the second sentence above.

adrienbernede replied:
I am using 2 parameters to name the directory `mfem_${CI_COMMIT_REF_SLUG}_${CI_PIPELINE_ID}`:
- CI_COMMIT_REF_SLUG is the ref name (converted), usually the branch/tag name
- CI_PIPELINE_ID
Since the pipeline ID is unique, ref name is not necessary, but it make the directory name more human readable: I can easily tell where the working dir for my branch is.

tzanio commented:
Maybe that's too much to ask, but can we have a GitLab badge like the Travis and Appveyor one at https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md?

tzanio commented:
LGTM

tzanio commented:
> LGTM

@adrienbernede -- I don't see a run for this at https://lc.llnl.gov/gitlab/mfem/mfem/pipelines... am I missing something?

tzanio commented:
Looking at this again, I wonder if it is better to merge as is with the `baseline` test only so the team can get used to running it, and add longer tests later?

adrienbernede commented:
@tzanio, I should have shared my changes on Friday...
I changed the way I manage the allocation to allow the addition of sample-runs-baseline. The consequence is that it is much more difficult to get resource on daytime. But the run-samples-baseline are here, and I just have to wait hours between each "validation" cycle.

I also added some comments in the yaml files, but plan on describing the pipeline more thoroughly. What is the best place to document it?

If you want to merge the baseline only version, I can add a commit doing the necessary, removing the sample-runs-baseline and moving back to debug queue for more responsiveness.

adrienbernede commented:
LGTM

adrienbernede commented:
@tzanio, this is working for me, however I’m not sure why I have differences in baseline test (except because of timing).

adrienbernede commented:
LGTM

tzanio commented:
LGTM

jandrej commented:
Merged in `next` for testing...

tzanio commented:
Adding a few extra days for approval...

adrienbernede commented:
@tzanio sorry for the forced push, I had a merge with master in my diverged history. I cherry-picked your changes. This time it was carefully done.

adrienbernede commented:
LGTM

adrienbernede commented:
Baseline and rebaseline have been extensively tested to check if it would work on master, and check if the commit works. I used a clone to emulate the behavior on master, and tested to rebaseline in another tests repository.

tzanio commented:
Adding a bit more time to review...

adrienbernede commented:
LGTM

tzanio commented:
Is this ready to go @adrienbernede and @acfisher ?

adrienbernede commented:
I am on a mirroring issue with this branch...

adrienbernede commented on .gitlab-ci.yml:
```diff
@@ -286,7 +283,7 @@ stages:
       fi
     - git add baseline-${SYS_TYPE}.saved
-    - git commit -m "${SYS_TYPE} rebaselined in Gitlab pipeline ${CI_PIPELINE_ID}"
+    - git commit -m "${SYS_TYPE} rebaselined in Gitab pipeline ${CI_PIPELINE_ID}"
```
```suggestion
    - git commit -m "${SYS_TYPE} rebaselined in GitLab pipeline ${CI_PIPELINE_ID}"
```

adrienbernede commented:
LGTM

acfisher commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,291 @@
+
+# The setup job in setup stage don’t rely on MFEM git repo. It prepares a
+# pipeline-wide working directory downloading/updating external repos.
+# TODO: update is not necessary anymore, repo are never shared with another
```
Is this todo still required?  I don't see an update step here.

adrienbernede replied:
The comment is not clear enough:
It’s saying that in the setup job, it’s not necessary to update git repositories, or even to test if they already exist, because for each pipeline, we create another ${BUILD_ROOT}. I would like to leave such a comment to remember that this is not optimal, so we you should eventually decide to optimize this (once it’s robust) or assume the extra amount of data generate.
In general, I think we will need to address the fact we will be accumulating a lot of data.

adrienbernede replied:
Made the comment more clear.

acfisher commented on .gitlab/quartz.yml:
```diff
@@ -0,0 +1,153 @@
+  extends: [.baselinecheck_mfem, .on_quartz]
+  needs: [allocate_quartz]
+
+  #samplebaselinecheck_mfem_gcc_6_1_0_quartz:
```
Remove dead code.

adrienbernede replied:
This is indeed dead, but waiting to be resurrected :)

adrienbernede replied:
Done

acfisher commented:
Is the mirroring up and running?  I don't see a pipeline launch under the latest magic word.

adrienbernede commented:
> Is the mirroring up and running? I don't see a pipeline launch under the latest magic word.

It’s not working, I am talking with @tgmachina about this.

tzanio commented:
LGTM

adrienbernede commented:
@tzanio No change since the last mirroring, so sing-off won’t have effect here.
Let me push the changes requested by @acfisher so that you can play with it.

tzanio commented:
LGTM

adrienbernede commented:
The PR is now fully functional from my perspective, let’s see how it scales!

adrienbernede commented:
Reminder: Sign-off does not "trigger" CI, but allows it, so that the next time mirroring happens, CI will be triggered.
Mirroring can be forced on CI side, which I did to speed-up the process here.

acfisher commented:
Looks like it is in good shape.  Lets get this pulled in asap.

v-dobrev commented on CONTRIBUTING.md:
```diff
@@ -370,6 +370,9 @@ Before you can start, you need a GitHub account, here are a few suggestions:
   error messages and the `tests/scripts` directory.
 
+- If triggered, track the status of the LLNL GitLab tests. If failing, ask
+  one of the _LLNL developer_ for details.
```
typo: developer -> developers

tzanio commented:
Merged in `next` for testing...

v-dobrev commented:
I "parmetis" build in the pipeline -- not sure why it is there -- we can probably remove it.

adrienbernede commented:
I think Tzanio mentioned it as well. I was assuming `MFEM_USE_MPI` implied `parmetis`. 
@acfisher Do you remember what was the intention?

v-dobrev commented:
Some TPLs depend on parmetis but mfem does not depend on it directly. The parallel mfem build requires just serial metis.

tzanio commented:
We can change that later... I think for now the goal is for folks to get experience with GitLab.
